### ISSUE TITLE
Consider automatically backing out profiler provided IL that triggers InvalidProgramException

### ISSUE THREAD

**Issue by noahfalk** [dotnet/coreclr#21778]
As seen in #18323, diagnosing an IPE caused by a profiler can be very challenging and time consuming for all involved. It is also makes applications less robust when a profiler fault instrumenting can easily transform into a failing application.

We should look into catching (or filtering) the IPE at runtime for profiler provided IL code bodies, followed by reactivating the original IL for the method (as if the profiler had called Revert) and reporting the failure via a non-fatal callback to the profiler. This provides the profiler the ability to log the error, potentially take corrective action, and then either continue the application with reverted IL or trigger its own fatal exception. If the profiler does elect to propagate an error it should be more obvious to everyone that the issue involves profiler instrumentation and is distinct from other root causes of an IPE.

**Comment by noahfalk** [comment]
cc @leculver @tommcdon @sywhang

**Comment by leculver** [comment]
I agree with the idea of backing out bad IL in principle, but I have two concerns:

1.  We can't be sure that profilers will function properly if a method that got unexpectedly rolled back mutated global state.  As a contrived example, a profiler may instrument Foo and Bar:  `Instrumented Foo` initializes a global variable (but `Instrumented Foo` gets rolled back and `Foo` runs instead).  Then, when the `instrumented Bar` runs (expecting `Instrumented Foo` to have set up global state) it hits an AV.
2.  Silently rolling back code means we have an _even more subtle bug_ to locate.  So if some profiler has a feature that instruments 9 methods successfully and 1 unsuccessfully, and that causes a non-crashing behavioral change that could take literally forever to realize that the feature isn't working correctly due to one small part of the whole feature not being enabled.  We don't have a way to inform the profiler at runtime that their method was rolled back.

This is mostly devil's advocate.  I think for 1 above the answer is "yeah but it'll crash less now, leading to better reliability anyway".  I think the mitigation for 2 is to set a global variable somewhere in CLR like `g_totalInstrumentedRollbacks`, or something similar that can be easily checked by a dev (and developer tools can be enlightened to check it).  Also, maybe consider some kind of callback or notification in the profiler API itself when a method is 'forcibly' rolled back.

**Comment by noahfalk** [comment]
Thanks for the feedback Lee!

> Also, maybe consider some kind of callback or notification in the profiler API itself when a method is 'forcibly' rolled back.

Yeah, that was already my intent "reactivating the original IL for the method (as if the profiler had called Revert) and reporting the failure via a non-fatal callback to the profiler."

**Comment by SergeyKanzhelev** [comment]
Is re-instrumenting of the method with the same `FunctionID` is good enough indication that something broke? Back in AVIcode days we kept original body reference and if we saw re-JIT we restored the body (or something like this).

One thing that can help is if `JitCompilationFinished` callback will return compilation result and you can force additional JIT checks for profiler-provided IL code. Like PEVerify on individual method. This alone may prevent most of the issues.

**Comment by noahfalk** [comment]
Thanks Sergey!

> Is re-instrumenting of the method with the same FunctionID is good enough indication that something broke?

Did you mean that the profiler would infer that the runtime had rolled the IL back because it is observing an additional set of Jit start/stop events without having queued a re-jit request itself? I think it would be error-prone because there are a number of other reasons the runtime would trigger the jit multiple times:
a) different generic instantiations of the same method
b) different optimization tiers of the same method
c) multiple threads that raced to jit the same method (one of the threads will ultimately lose the race and its code will never be used, but that doesn't prevent the profiler from observing the JIT happened)

> One thing that can help is if JitCompilationFinished callback will return compilation result

I think that could be on the table. As I recall current behavior is that the IPE causes execution to skip the JitCompilationFinished entirely, so right now the profiler gets start but no end. Given that JitCompilationFinished might already have some behavior attached to it in the profiler I'd be leary to change that by default, but if this wound up as an opt-in then we could be more certain that the profiler had coded the callback appropriately.

> can force additional JIT checks for profiler-provided IL code. Like PEVerify on individual method. This alone may prevent most of the issues.

Are you talking about doing this online or offline? Online if a profiler discovered the IL it generated was bad I wasn't expecting it would be able to recover beyond logging an issue for offline analysis or disabling some portion of functionality. In an offline capacity I assume profiler vendors are already testing their generated IL and if those in-house tests failed they would investigate. I imagine better diagnostics about why the JIT failed would be appreciated to make that offline edit, test, debug cycle more expedient and I'm happy if you want to have a separate issue with a request like that?

**Comment by SergeyKanzhelev** [comment]
> Did you mean that the profiler would infer that the runtime had rolled the IL back because it is observing an additional set of Jit start/stop events without having queued a re-jit request itself?

Yes, we implemented something like this with some additional heuristics to protect from "a" and "c". My point is that with an additional details CLR provides - profiler theoretically can recover by itself.

> Are you talking about doing this online or offline? Online if a profiler discovered the IL it generated was bad I wasn't expecting it would be able to recover beyond logging 

I meant online. My understanding is that JIT compilation uses a lot of heuristics and not necessarily checks the validity of IL code as long as it was able to generate machine code. I suggest that more thorough validation enablement for profiler-provided code may be a good option to minimize number of IPE.

**Comment by noahfalk** [comment]
> Yes, we implemented something like this with some additional heuristics to protect from "a" and "c".

I can imagine its possible and nothing would prevent a profiler from doing it, but I imagine if you had the option between building these heuristics or having the runtime explicitly tell the profiler that the IL was invalid and the edit was rolled back it would be much easier to consume the latter.

> My understanding is that JIT compilation uses a lot of heuristics and not necessarily checks the validity of IL code as long as it was able to generate machine code. I suggest that more thorough validation enablement for profiler-provided code may be a good option to minimize number of IPE.

I think you are suggesting that we avoid IPE getting thrown in the JIT by writing a pre-JIT validation phase that will detect all the invalid IL cases that would have caused an IPE and then do some other error behavior instead? Technically we could do this, but writing an independent pre-verification algorithm is expensive both in developer time and runtime performance. I am suggesting we catch the IPE thrown by the JIT so that we can leverage all the implicit validation it already does. From the profiler's perspective both implementations prevent the application from crashing with an unhandled IPE and provide a callback indicating the IL was invalid so I don't think it would matter to the profiler which technique we used? Let me know if you think I am overlooking something.

Increasing the strictness of validation checking, whether inside or outside the JIT, would have negative runtime perf and back-compatibility impact so its not a path I'd want to pursue. If a profiler wanted to opt-in to stricter checking of only its own code I'm not opposed to that on principle, but I don't think it would be cheap to implement and I suspect it would catch extremely few additional issues not already caught by the JIT. Its not something I plan to address with this issue but I'm fine if you wanted to open another issue about it.

**Comment by noahfalk** [comment]
> but it is also pandoras box with respect to "what is the right behavior after something is backed out?

The profiler certainly has the option to throw a fatal exception of their own. That wouldn't lessen the number of crashes if they chose that route, but it could improve diagnosability assuming the profiler logged something about what method had the bad IL and the situation that caused them to generate it.

> what state the reverted IL was in

As a strawman I propose if you continue from the error callback then the IL is in the state that you would get if you had called Revert().

> I think taking @SergeyKanzhelev suggestion of having a validation path and merging it with the idea of a API call that says "heavily validate this IL" but also leaving the fast path might be the best approach

The mechanism we have today to validate IL is to JIT it. Is the issue that its hard to ensure code gets jitted at development time so we should add some API that forces a method to get jitted even if managed code hasn't called it? If you are proposing there should be a validation that explicitly doesn't JIT I would want to know why it mattered : )

Thanks!

**Comment by AaronRobinsonMSFT** [comment]
> The profiler certainly has the option to throw a fatal exception of their own.

Agreed, but this is a discussion about _not_ coming down hard and that is my point. When an operation for instrumentation is performed, there is an expectation - at least in all the time I was on the profiler team - that the supplied code/IL/assembly was valid and wasn't going to terminate/corrupt/etc the process. Going down the route of trying to accept code and if it isn't okay backing out the request seems like the runtime doing more work than it should. Someone on the profiler implementer side made a mistake and at that point there are bound to be invariants in the profiler anyways - assuming they meant for the instrumented code to perform some action.

>so we should add some API that forces a method to get jitted even if managed code hasn't called it

Not really, it was merely a concrete example of a contract and observation about the underlying issue. Which is if profiler writers are supplying IL that is bad and can cause an IPE, then _that_ is the bug and something should be done to help make generating custom IL easier and less error prone. If that is an API that performs some validating at debug/development time so be it, but again the issue is not bad IL can be supplied it is that it is being supplied, so I would argue maybe we should consider the scenarios that is happening and how we can help with that. 

I can almost guarantee that as soon as bad IL is provided and then returned, the next question that we will need to answer is "what is wrong with it"? Telling people something is wrong without "why" it is wrong and how to move past it is a consistent problem in this space and having personally dealt with the ubiquitous `E_FAIL` that seems to be the only error code from a lot of the diagnostic interfaces gets tiresome.

**Comment by leculver** [comment]
> When an operation for instrumentation is performed, there is an expectation - at least in all the time I was on the profiler team - that the supplied code/IL/assembly was valid and wasn't going to terminate/corrupt/etc the process. Going down the route of trying to accept code and if it isn't okay backing out the request seems like the runtime doing more work than it should. 

@AaronRobinsonMSFT In a perfect world, yes we expect that profiler developers are giving us valid, well-formed (and functional) IL and we wouldn't have to do extra work when a profiler gets it wrong.  However, the reason this issue is being raised is that we **did** have that as a policy, and it harmed our customers.

We spent 5 months trying to track down the issue and even after we got a really consistent set of crash dumps to look at it _still_ took 4 developers over a month to track down.  (And to be clear: that's 4 people who really know what they are doing in this space...it could have taken a lot longer).  That's really strong evidence we should do something here, if we can come up with a design that would help.

> I can almost guarantee that as soon as bad IL is provided and then returned, the next question that we will need to answer is "what is wrong with it"?

I mean sure, we could always try to make this feature better and better.  That's not the point here.  The issue/feature that Noah has raised here solves a very concrete problem with measurable customer impact:  We can get into a state where a profiler can cause seemingly random failures in the runtime that we have no real mechanism to track down.  We have no mechanism to inform the profiler that they made a mistake.  We have no mechanism to help our customers out that run into this issue.

Instead our end users are faced with a crashing application/service that there is no mechanism to diagnose what's gone wrong.  Noah's proposal fixes that by ensuring (mostly) that their service will function in production, and in the cases where it won't work properly (my second point above), we will provide callbacks and diagnosability so it doesn't take 4 Microsoft developers months to solve.

If we found a reasonable way to fully validate IL while we are at it, great!  But that's not solving the [real customer issue](https://github.com/dotnet/coreclr/issues/18323) that this proposal is aiming to fix.

**Comment by leculver** [comment]
> based on your description the issue here is "what happened?" rather than "the runtime needs to soldier on regardless of what a loaded native component did".

No, we need to do both, if possible.  Give better diagnostics _and_ be more resilient to bad edits.  We can walk and chew bubble gum here.

> Admittedly the latter case will cause a more obvious stack or failure, but so would providing richer details in IPEs. No?

Ignoring everything else, I'm happy to have more details in an IPE.  That's a fine goal to have, and you have a +1 from me in any work that gives better diagnostics in that space.

However, I still think you are fundamentally missing the point:  Let's say we fully extended InvalidProgramExceptions to include everything you mention and more.  Now let's say a profiler developer (as we had in the linked thread) makes a mistake and poorly instruments a method.  When an application developer who has deployed a service to Azure, they now see their application is failing with InvalidProgramExceptions.  With this new feature they will know that Foo method failed, and here's the IL that's bad, and so on...  They still won't have any idea why this is happening, and will still need our help to solve it.

(Validating IL is another can of worms, as Noah points out.)

**Comment by noahfalk** [comment]
I think we do have two separate tracks, and we can evaluate them independently:

1) The runtime could revert IL to make forward progress. I don't want the runtime to be doing that unilaterally, the profiler needs to tell us somehow that rolling back the IL is OK. In my initial suggestion that meant 'not throwing a fatal error' but I don't think the exact mechanism is important as long as we think the profiler writer understands their role. If most profilers are never going to be in a position to say that it is OK to continue with reverted IL then we probably shouldn't bother with this.

2) Improve the diagnostics - not what this issue was originally tracking but its in the conversation and we can always open a new issue for it to move ahead:
   a) Improving customer-visible diagnostics. If we crash with an IPE whose message said "IL for method bla was bad, the IL was modified by the XYZ profiler dll. Consider contacting the profiler vendor or disabling the profiler to resolve this issue" that would give the customer a way forward. 
  b) Diagnostics for profiler vendors. There is plenty of room for improvement here - what IL was bad, what method was it for, why was the IL bad, what was the original IL. 

My guess is that everybody likes more diagnostics, we just have to prioritize improvements there vs. all the other work we could be doing. Some parts are probably cheap, other parts (saying why the IL is bad) is probably expensive.
